### PR TITLE
[CR] Replace usage of `getch` with input manager

### PR DESCRIPTION
**Pull request by BevapDin**
Gets rid of direct usage of `getch`, and uses `input_manager::get_input_event` instead. In the future it should be replaced with an `input_context`.

The input manager calls `getch` in ncurses builds, but calls native functions in SDL and in Windows curses builds.

It also makes `input_manager::get_input_event` simpler: instead of one implementation for all platforms, it is split into several implementations, one for each platform.

--

I can't currently test the SDL nor the Windows code, so that needs to be done and verified by somebody else (therefor the CR).

--

If this gets in, I have a branch with a macro system (similar to that of Dwarf Fortress) ready, but it depends on this one.

### CODE REVIEW

**Comment by codemime** [comment]
Glad to see you back!

File: src/character.cpp [hunk 1]
```diff
@@ -1826,7 +1826,8 @@ hp_part Character::body_window( const std::string &menu_header,
     char ch;
     hp_part healed_part = num_hp_parts;
     do {
-        ch = getch();
+        // TODO: use input context
+        ch = inp_mngr.get_input_event(nullptr).get_first_input();
```
Comment (codemime) [inline src/character.cpp]: Why don't we make `nullptr` default in the definition?

File: src/input.cpp [hunk 2]
```diff
@@ -1096,14 +1096,28 @@ long input_manager::get_previously_pressed_key() const
...
-#ifndef TILES
+void input_manager::wait_for_any_key()
+{
+    while( true ) {
+        switch( inp_mngr.get_input_event( nullptr ).type ) {
+            case CATA_INPUT_KEYBOARD:
+                return;
+            // errors are accepted as well to avoid an infinite loop
+            case CATA_INPUT_ERROR:
+                return;
```
Comment (codemime) [inline src/input.cpp]: Could be a bit shorter with a fallthroug.

File: src/character.cpp (hunk 1 above)
Comment (BevapDin) [inline src/character.cpp]: I shall add an overload for this. And I'm glad to be back as well.

File: src/input.cpp [hunk 3]
```diff
@@ -1096,14 +1096,33 @@ long input_manager::get_previously_pressed_key() const
...
+            default:
+                break;
+        }
+    }
+}
+
+input_event input_manager::get_input_event()
+{
+	return get_input_event( nullptr );
```
Comment (BrettDong) [inline src/input.cpp]: Should use spaces to indent here.

Comment (BevapDin) [inline src/input.cpp]: Fixed.

**Comment by kevingranade** [comment]
> If this gets in, I have a branch with a macro system (similar to that of Dwarf Fortress) ready, but it depends on this one.

I'd need a really strong argument for why a macro system is needed.  It seems to me that macro systems are generally used to paper over bad interfaces that require excessive repetition, we should be fixing any such issues directly instead.

**Comment by BevapDin** [comment]
> I'd need a really strong argument for why a macro system is needed.

I use it currently for repairing with the welder (character is masochist). I need 4 key strokes: `e`xamine, `3` (direction), use`w`elder, `2` (repair, `1` would be cauterize for fun).

Those 4 strokes are in the macro, now I only press `F2` (play macro), select the item and press `3` to repair until fully repaired, and later `q` to quit the repair menu.

I also use it for firing arrows while I have different types of arrows available: `f`ire, `c` (invlet of arrows to use), `a`im and fire.

**Comment by kevingranade** [comment]
On Jan 9, 2017 8:06 AM, "BevapDin" <notifications@github.com> wrote:

I'd need a really strong argument for why a macro system is needed.

I use it currently for repairing with the welder (character is masochist).
I need 4 key strokes: examine, 3 (direction), usewelder, 2 (repair, 1 would
be cauterize for fun).

Those 4 strokes are in the macro, now I only press F2 (play macro), select
the item and press 3 to repair until fully repaired, and later q to quit
the repair menu.

Sounds like we should just have a "keep repairing" type action.

I also use it for firing arrows while I have different types of arrows
available: fire, c (invlet of arrows to use), aim and fire.

If I understand correctly, you have a different macro for each stack of
arrows?
That's not as simple, but I still feel like we need to solve this problem
for everyone, not just people who know how to define macros.

**Comment by BevapDin** [comment]
> Sounds like we should just have a "keep repairing" type action.

So you want to implement *this* macro in C++? What about the next process that takes multiple key strokes, will this be implemented in C++ as well?

> If I understand correctly, you have a different macro for each stack of
> arrows?

No, just for the stack I'm using. I have not yet implemented saving and loading of macros. It replaces 3 key strokes with one, a "performance" gain of 66%! (-; Yes, I'm lazy.

> That's not as simple, but I still feel like we need to solve this problem
> for everyone, not just people who know how to define macros.

Macros are simple: you press one button to start recording, do whatever you want, press the button again to stop recording and press another button to play the macro. As often as you want.

Another example, planting seeds with a hoe:
`A` to apply the wielded hoe (or `a` and the invlet of the hoe),
`e`xamine the `.` ground to plant `2` (the second seed type in the inventory),
`4` to move to the next tile to the west.

You can repeat this 19 times (pressing the play button 19 times), planting 20 seeds in an orderly row.
A similar macro can be used to harvest it all (examine, yes to query, pickup fruits).

Another example: cleaning rubble:
press `e`xamine, press `1` (SW), press `ESC` to exit the pickup menu, repeat for S, repeat for SE (cleans the 3 tiles to the south), press `2` to move south. Repeat as long as needed. Can even be extended to include picking up all the items (press `,` and `ENTER` instead of `ESC`).

--

The actions that a macro system automates vary too much to implement them in C++, a macro system is intuitive (recording and repeating key strokes) and is flexible.

File: src/wincurse.cpp [hunk 4]
```diff
@@ -591,9 +591,7 @@ uint64_t GetPerfCount(){
     return Count;
 }
 
-//Not terribly sure how this function is suppose to work,
-//but jday helped to figure most of it out
-int curses_getch(WINDOW* win)
+input_event input_manager::get_input_event( WINDOW *win )
```
Comment (kevingranade) [inline src/wincurse.cpp]: Fogot to #include "input.h" in tis file, I'll fix up in the merge unless something else breaks.

File: src/wincurse.cpp [hunk 5]
```diff
@@ -627,9 +625,40 @@ int curses_getch(WINDOW* win)
...
+        } else {
+            rval.type = CATA_INPUT_ERROR;
+        }
+    } else {
+        if( lastchar == 127 ) { // == Unicode DELETE
+            previously_pressed_key = KEY_BACKSPACE;
+            return input_event( KEY_BACKSPACE, CATA_INPUT_KEYBOARD );
+        }
+        rval.type = CATA_INPUT_KEYBOARD;
+        rval.text = UTF32_to_UTF8( lastchar );
```
Comment (kevingranade) [inline src/wincurse.cpp]: fixing in merge, utf32_to_utf8(), probably left caps lock on after typing previous line or something.  Also #include "catacharset.h".

**Comment by kevingranade** [comment]
Did a windows build, everything looks good there, I'm doing a SDL build now.

**Pull request merged**